### PR TITLE
Sanitize the domain of the URL for Telemetry

### PR DESCRIPTION
**Pull request by Venkatesh-rengasamy**
### :nut_and_bolt: Description: What code changed, and why?

Sanitize the domain URL in request payload for Telemetry event calls and Segment api calls. 

### :chains: Related Resources

#5237

### :+1: Definition of Done

Sanitized the domain URL for Telemetry event calls and Segment api calls. 

Telemetry event api for page type -> sends referrer & url value as empty string in request payload.

Segment api's request payload contains -> properties.referrer & url as empty string And context.page.referrer & url values are masked.

FYI - Segment api calls as follows:-
https://api.segment.io/v1/p
https://api.segment.io/v1/t
https://api.segment.io/v1/i
https://api.segment.io/v1/g

### :athletic_shoe: How to Build and Test the Change

Enable the telemetry checkbox in About popup. Reload the page and open the network tab. Check the referrer & url values in request payload of Telemetry event api and segment api calls.  

### :white_check_mark: Checklist

**All PRs** must tick these:

- [x] I have read the [CONTRIBUTING document](https://github.com/chef/automate/blob/master/CONTRIBUTING.md).
- [x] All commits signed-off for the [Developer Certification of Origin](https://github.com/chef/chef/blob/master/CONTRIBUTING.md#developer-certification-of-origin-dco).

With occasional exceptions, all PRs **from Progress employees** must tick these:

- [x] Is the code clear? *(complicated code or lots of comments--subdivide and use well-named methods, meaningful variable names, etc.)*
- [x] Consistency checked? *(user notifications, user prompts, visual patterns, code patterns, variable names)*
- [x] Repeated code blocks eliminated? *(adapt and reuse existing components, blocks, functions, etc.)*
- [x] Spelling, grammar, typos checked? *(at a minimum use `make spell` in any component directory)*
- [x] Code well-formatted? *(indents, line breaks, etc. improve rather than hinder readability)*

All PRs **from Progress employees** should tick these if appropriate:

- [ ] Tests added/updated? (all new code needs new tests)
- [ ] Docs added/updated? (all customer-facing changes)

*Please add a note next to any checkbox above if you are NOT ticking it.*

### CODE REVIEW

File: components/automate-ui/src/app/services/telemetry/telemetry.service.ts [hunk 1]
```diff
@@ -233,11 +237,21 @@ export class TelemetryService {
     );
   }
 
-  private emitToPipeline(operation: String, payload: Object) {
+  private emitToPipeline(operation: String, payload: any) {
```
Comment (vinay033) [inline components/automate-ui/src/app/services/telemetry/telemetry.service.ts:240]: can we define a specific type for payload?

File: components/automate-ui/src/app/services/telemetry/telemetry.service.ts [hunk 2]
```diff
@@ -273,4 +287,38 @@ export class TelemetryService {
...
+    if (payload && payload.obj && payload.obj.context && payload.obj.context.page) {
+      const page = payload.obj.context.page;
+      if (page.referrer) {
+        page.referrer = analytics.sanitizeDomainURL(page.referrer);
+      }
+      if (page.url) {
+        page.url = analytics.sanitizeDomainURL(page.url);
+        console.log(page.url);
+      }
+      if (payload.obj.context.ip) {
```
[the remark this answers is not in the document]
Comment (kalroy) [reply components/automate-ui/src/app/services/telemetry/telemetry.service.ts:317]: Checked the upstream pipeline and the it is not getting masked.

**Review by kalroy — CHANGES_REQUESTED** [review]

File: components/automate-ui/src/app/services/telemetry/telemetry.service.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (kalroy) [reply components/automate-ui/src/app/services/telemetry/telemetry.service.ts:317]: Raise another issue for further investigation.

File: components/automate-ui/src/app/services/telemetry/telemetry.service.ts [hunk 3]
```diff
@@ -273,4 +287,38 @@ export class TelemetryService {
...
+
+  middleware({ payload, next }) {
+    if (payload && payload.obj && payload.obj.context && payload.obj.context.page) {
+      const page = payload.obj.context.page;
+      if (page.referrer) {
+        page.referrer = analytics.sanitizeDomainURL(page.referrer);
+      }
+      if (page.url) {
+        page.url = analytics.sanitizeDomainURL(page.url);
+        console.log(page.url);
```
[the remark this answers is not in the document]
Comment (Venkatesh-rengasamy) [reply components/automate-ui/src/app/services/telemetry/telemetry.service.ts]: console logs are removed.

File: components/automate-ui/src/app/services/telemetry/telemetry.service.ts (hunk 1 above)
Comment (Venkatesh-rengasamy) [reply components/automate-ui/src/app/services/telemetry/telemetry.service.ts]: payload param is common to accept all types of events from segment. Each event have different set of properties. I removed 'any' type change and keep previous code.

File: components/automate-ui/src/app/services/telemetry/telemetry.service.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (Venkatesh-rengasamy) [reply components/automate-ui/src/app/services/telemetry/telemetry.service.ts]: Reverted the IP changes and created separate issue for it. #5575

File: components/automate-ui/src/app/services/telemetry/telemetry.service.ts [hunk 4]
```diff
@@ -273,4 +285,34 @@ export class TelemetryService {
   private getCurrentDateTime() {
     return (new Date).toISOString();
   }
+
+  sanitizeDomainURL(url) {
```
Comment (vinay033) [inline components/automate-ui/src/app/services/telemetry/telemetry.service.ts:289]: please define type here

File: components/automate-ui/src/app/services/telemetry/telemetry.service.ts [hunk 5]
```diff
@@ -273,4 +285,34 @@ export class TelemetryService {
...
+    let restByDot: any;
```
Comment (vinay033) [inline components/automate-ui/src/app/services/telemetry/telemetry.service.ts:290]: also, please do not use `any`.

File: components/automate-ui/src/app/services/telemetry/telemetry.service.ts [hunk 6]
```diff
@@ -273,4 +285,34 @@ export class TelemetryService {
...
+      restBySlash = restBySlash.join('/');
+      firstBySlash = '***';
+      return firstByDot + '.' + firstBySlash + '/' + restBySlash;
+    } else {
+      restByDot = '***';
+      return firstByDot + '.' + restByDot;
+    }
+  }
+
+  middleware({ payload, next }) {
```
Comment (vinay033) [inline components/automate-ui/src/app/services/telemetry/telemetry.service.ts:305]: same for payload and next

**Review by kalroy — APPROVED** [review]

**Pull request merged**